design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,162 @@
+
+#### Objects Path Conventions
+
+Uncommitted GC must scan the bucket in order to find objects that are not referenced by lakeFS.
+To optimize this process, suggest the following changes:
+
+1. Divide the repository namespace into time-and-size based partitions.
+2. Partition name will be a time based, reverse sorted unique identifier.
+3. LakeFS will create a new partition on a timely basis (for example: hourly) or when it has written < TBD > objects to the partition.
+4. Each partition will be written by a single lakeFS instance in order to track partition size.
+5. The sorted partitions will enable partial scans of the bucket when running the optimized GC.
+
+#### StageObject
+
+The StageObject operation will only be allowed on addresses outside the repository's storage namespace. This way, objects added using this operation are never collected by GC.
+
+#### [Get/Link]PhysicalAddress 
+
+1. GetPhysicalAddress to return a validation token along with the address (or embedded as part of the address).
+2. The token will be valid for a specified amount of time and for a single use.
+3. lakeFS will need to track issued tokens/addresses, and delete them when tokens are expired/used
+4. LinkPhysicalAddress to verify token valid before creating an entry.
+    1. Doing so will allow us to use this time interval to filter objects that might have been uploaded and waiting for
+       the link API and avoid them being deleted by the GC process.
+    2. Objects that were uploaded to a physical address issued by the API and were not linked before the token expired will
+       eventually be deleted by the GC job.
+>**Note:** These changes also solves an existing bug where an issued physical address which was committed can be overwritten by Link
+
+#### S3 Gateway CopyObject
+
+1. Copy object in the same branch will work the same - creating a new staging entry using the existing entry information.
+2. For objects that are not part of the branch, use the underlying adapter copy operation.
+
+#### Move/RenameObject
+Clients working through the S3 Gateway can use the CopyObject + DeleteObject to perform a Rename or Move operation.
+For clients using the OpenAPI this could have been done using StageObject + DeleteObject.
+To continue support of this operation, introduce a new API to rename an object which will be scoped to a single branch.
+
+#### PrepareUncommittedForGC
+
+A new API which will create meta-ranges and ranges for a given branch using its uncommitted data. These files
+will be saved in a designated path used by the GC client to list branch's uncommitted objects.  
+For the purpose of this document we'll call this the `BranchUncommittedMetarange`
+
+### GC Flows
+
+The following describe the GC process run flows on a repository:
+
+#### Flow 1: Clean Run
+
+1. Listing namespace objects
+   1. Read all objects directly from object store (can be done in parallel using the partitions) -> `Store DF`
```

Comment:
This is a really cool idea.  But we should detail a bit more how we parallelize this: since storage partitions are time-based, it will be quite hard to know how many parallel tasks to generate, and also quite hard to ensure that each parallel task is large enough.  The issue is that I do not know how many objects are created every minute, so I have no idea how to select start points for each task.

One way to do it might be to use _only_ size-based storage partitions, and then add an API call that gives the estimated size of each storage partition and the first and last storage partitions.  That makes parallelizing the read in the Spark client easier.

Reply:
It will be very difficult to align the data we're supposed to read from the namespace, with the commits we read without making the slices time-based. We added a upper bound to the slice size so that we can efficiently divide the work.